### ISSUE TITLE
Add the ability to load pages based on an API response, rather than by URL

### ISSUE THREAD

**Issue by dan-hammond** [nuxt/nuxt#7437]
### What problem does this feature solve?

# TL;DR
When a user visits the site, I want to be able to call an API that returns content from my CMS, and then load the relevant page component based on that response. 

I do not want to sacrifice the Nuxt features like asyncData or layouts in the process.

An example might look like this:

1) User visits /case-studies/case-study-one
2) Nuxt calls our CMS API with the above slug
3) Nuxt gets a response from the API with the relevant content and the page template to use (for this example, `{ template: "caseStudy" }`)
4) Nuxt routes the user to `~/pages/caseStudy.vue`

I would like this functionality to be installable as a module for use in many sites.

# Background
I am trying to use Nuxt (in Universal mode) with a CMS that is very flexible about the URLs for content. It is not possible for us to know what kind of content we're trying to access based on the URL. Because of this, we're currently struggling to use Nuxt with our CMS, as it is very strict about its URLs and routing.

Below are some examples of situations that can occur, which would break or otherwise not be possible with the current routing. 

In most cases it's likely possible to suggest one-off workarounds, but fundamentally those workarounds won't provide the level of flexibility that our CMS users have come to expect. 

Any proposed workarounds are likely to be fine for one project, but we're keen to use Nuxt to produce many websites, so the overhead of managing these restrictions becomes less tenable.

## Example 1 
We create a `/products` route in Nuxt, which fetches content for the `products` page in the CMS. 

A user of the CMS then decides they want to rename this page, as they only sell books, so they call it `books`. This is something that CMS users are free to do at present.

The Nuxt routes no longer reflect what is set in the CMS -- a developer would have to be involved to release a new version of the front-end, so that URLs are properly updated to say `books`. 

In the worst case scenario, the API call could fail if the CMS does not properly redirect requests from `products` to `books` following the rename. Our CMS does keep track of such renames, but it is a point of fragility that should be noted.

## Example 2 
In the CMS we create a hierarchical set of content that we'll call product categories, which can be infinitely nested. It is possible to put products in any category at any level. As such, we cannot tell whether a page is a product or a category based on its URL.

`/products/category-1/collection-1` -- is this a product that includes the collection of all products from subcategories? Or is it a subcategory?

Currently, Nuxt would require us to flatten our URLs by moving the products to a separate, flat folder in the CMS. This is a terrible user experience for users working in the CMS, especially if there are hundreds or even thousands of products.

The above scenario can also occur with other hierarchical content that can't easily be flattened, such as microsites. It's not reasonable to ask CMS users to put all microsite pages from all microsites into one generic folder, as it becomes a disorganised mess.

## Example 3
The website has a plethora of content pages at the top level, such as `/about-us`, `/cookie-policy`, `/privacy-policy`, and so on. As the names and quantity of these are unknown, they are likely to be registered as a dynamic route. 

Our CMS users then want to create a short, memorable URL for a product that they're going to be heavily marketing, say `/product-one`. Currently this wouldn't work in Nuxt without some sort of special case to indicate that this URL is actually a product, so we should redirect to the product's normal URL.

## Example 4
Our CMS tracks pages that have moved, and redirects the user to the new URL automatically if they visit the old one. Currently, this is not something that we'd easily be able to do with Nuxt. 

Our API would need to give different responses for moved content, which Nuxt then picks up, and redirects to the new page, which will then need a new API call to fetch the content again. Ideally, we should only ever need to call our CMS' API once.

# Why are we trying to use Nuxt?
We're definitely keen to modernise our front-end technology by using Nuxt, rather than continue with our current approach of using server side technology.

- We have a small team of devs that work on a few projects, including SPAs made purely in Vue, so one language + framework for all projects reduces the amount of skills we need.

- Nuxt and Vue are a joy to use, and so we want to use it for these projects too.
  
- A lot of optimisation is taken care of for us, rather than us having to figure out how to keep up with the latest optimisation techniques in our comparatively dated CMS/backend.

- Some of the pages we work on have complex, interactive components that are nightmarish without a front-end framework (and we want to go further down this road.)

- We get some niceties like page transitions, easy animations, and so on.

- Because we build a lot of sites, we try to pack up anything reusable in private packages. It's very hard to make reusable front-end components with the CMS we're using, especially keeping customisation + optimisation in mind.

We have also used Vue with our CMS to produce a few websites, by creating a separate Vue app for each template in the CMS (I believe this is acronymed as MPA in the Vue world). This worked for us, but it's definitely rough around the edges compared to going all in with an SPA and using the CMS in a fully headless role. 

A few of the problems we've encountered by keeping each page a separate app:

- Optimisation and code reuse is not a pleasant experience.

- We had to maintain Webpack configs ourselves, which is also unpleasant.

- Providing content from the CMS into the Vue app involved mixing two templating languages (to avoid SEO and further performance issues.)

- Some features we want to provide involve multiple pages within the same SPA, and things like routing start to get confused.

### What does the proposed changes look like?

Hopefully this has been partially explained above. Ideally a new module would be created that allows an async function to be defined somewhere, which is called prior to any routing takes place when attempting to visit any page. This async function would then be able to provide the relevant page component to load, and also commit the content to Vuex, or pass the data through as props to the component. 

It is important to note that this is not the same as redirecting to pages, as that would change the destination URL. This is more about delegating the routing to the API (or indeed any other user specified code), so that the correct page can be loaded.

Whilst I appreciate that this seems to be a niche use case, I think it's likely that more people are going to want something along these lines as the popularity of headless CMS solutions increases. I think it would provide a huge amount of flexibility around routing that doesn't presently exist, and I can see the functionality being used in other ways.

My colleagues and I have spent over a week trying to implement this ourselves as a plugin or module, but we're getting nowhere fast.

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c10711">#c10711</a>)</em></sub></div>

**Comment by sugoidesune** [comment]
I see your point.
You could try to use dynamic components.
https://vuejs.org/v2/guide/components-dynamic-async.html

This would allow you to render a completely different interface dynamically.
The modularity would also allow to structure it cleanly.

**Comment by RCVarley** [comment]
I've been trying to crack this problem for a couple of weeks now, as I was sure that this behaviour was supported, and that I just didn't know how to get it working. 

> Ideally a new module would be created that allows an async function to be defined somewhere, which is called prior to any routing takes place when attempting to visit any page.

My intended solution, much like you're describing, was to write a plugin with the following:

```js
await context.app.router.beforeEach( async ( to, from, next ) => {
    /** 
     * Prepend the cms provided url's path with /content,
     * which the API knows is for CMS content
     */
    const content = await context.app.$axios.$get( `/content${context.route.path}` );
    
    // Save the resulting content to vuex
    context.store.commit('cms/load', content);

    /**
     * The content has a contentType property which is a string
     * containing the name of the component we want to load
     * 
     * This finds the corresponding route 
     */
    const routes = context.app.router.options.routes;
    const targetRoute = routes.find( r => r.name === content.contentType );

    // Sets the component to load as /Pages/[componentName].vue
    to.matched[0].components = {
      default: targetRoute.component
    };

    next();
  })
```

Unfortunately, this didn't work as, unlike middleware, plugins don't support asynchronous calls.  I figured that this was a strange discrepancy, so I reported is as a bug (#7429).

@sugoidesune dynamically loading components would work, but we'd lose a lot of the benefits of using nuxt.  The goal here is to let nuxt take care of everything other than which component to load.  That way we get all of the tasty nuxt features for free.  Plus, the implementation becomes a lot easier to maintain further down the line if it uses nuxt properly (as much as it can do).

**Comment by dan-hammond** [comment]
@sugoidesune Thank you for the reply. 

I've just spent the morning creating a page/wrapper component that will then use dynamic components to load the relevant pages. It does work, I am not getting SSR errors, and I do see the relevant components being loaded. It's definitely working better than my previous attempts at fiddling with the routing.

However, there are two fairly big caveats to this approach:

- Essentially all 'page' level functionality is now broken/inaccessible, as the components are no longer pages. The only things that I believe still work are 'head' and 'fetch'.

- Code splitting will no longer nicely split each page/route component into separate chunks, as I have to rely on a single dynamic import with string interpolation to load the relevant component. I can't specify different chunk names for different pages from this wrapper component.

I think it's likely that the latter problem could be solved by fiddling with some creative Webpack config, but I don't think there's a solution for the first point. 

Do you/does anyone have any ideas on how those issues could be tackled?

Thanks

**Comment by sugoidesune** [comment]
Im not familiar with the source code, so I don't know if a pull request or a workaround would be more feasable. Certainly any workaround that deveates too much would turn ugly rather quick.

@RCVarley why would a middleware be insufficient for fetching data and routing to the correct page?

@dan-hammond Re the first point, you could set those page specific functions from a store. 
each lets call it dynamic-page-component would be associated with store settings for page specific actions. its getting uglier by the minute haha

**Comment by dan-hammond** [comment]
@sugoidesune That's a good idea, definitely something to investigate. As you said though, it's getting rather ugly!

It looks like 'layout' can be a function which has access to the context -- so it could likely work there, but there are a few other options on pages which are only properties. Those might be more tricky.

The other big issue, of course, is code-splitting. Currently all of our page components will end up in one mega chunk, as there's no way (as far as I can see) to have Webpack put them all in different chunks.

Thanks again for your response, I really appreciate it.

**Comment by dan-hammond** [comment]
Thank you for investigating, and apologies for the delay in responding. I've had to park my work on this part of the project for the moment, as I'm making precious little progress, and have a looming deadline.

As soon as I've got other parts of the project in a more complete state, I'll go back to this and test out your suggestion.

Thanks once again.

**Comment by daltonrooney** [comment]
I'm very interested in your solution to this. I'm also using a CMS with a flexible routing structure (Craft CMS w/ GraphQL) and a page can be at any URI and can be one of many types... but we don't know which until we query the API. So I am loading components conditionally from `_.vue` depending on the typeHandle. No asyncData in those sub-components, though, which breaks a lot of things.

**Comment by dan-hammond** [comment]
I've not had a chance to really work on this at all since I last wrote a comment. I briefly tried out @sugoidesune's suggestion (thank you again btw), but I found that it didn't work when rendering on the server, as it's happening too late into the lifecycle.

I found a diagram that suggested that the `nuxtServerInit` action in Vuex is the earliest step of the lifecycle, so I tried it in there, to no avail. At that point it seemed to have no effect on routing either.

For the time being, I went back to the approach of using _.vue (well, a module equivalent of it) as a wrapper component which makes the API call and loads the relevant page using `<component :is>`. 

In my "page" components, I used `fetch` instead of `asyncData` as it seems to do much the same job. I also used  `nuxtServerInit` to update a value in my Vuex store with my desired layout, which I then use to set _.vue's `layout` property. I'm hoping to extend this approach to the other dynamic options, like `transition`.

I haven't found a way to pack the `nuxtServerInit` action up in a module or npm package, which is a shame. I've ended up with two Vuex files that I seemingly am unable to turn into a Nuxt module (one for `nuxtServerInit`, and a Vuex module for collecting the API response). I would've preferred to get all the Nuxt functionality working and hide the quirkiness away from the website's 'src' folder.

I hope the above is vaguely helpful, and that perhaps someone reading this manages to take what I've found so far and get a little further along with it.

**Comment by gianpesto** [comment]
Hi! I am happy to see i am not the only one with this problem.
I am having a very similiar construct with Nuxt. 

You need to log in as a User to access the page, after you successfully loged in, you are granted privileges (different for each user), this also affects the pages the user has access to.

So there is a navigation endpoint that returns a JSON with all paths that the user can navigate to. This JSON however is maintained through a CMS, so the paths might change according to the language.

So consider these files:

```
- cart
    index.vue
    checkout.vue
    payment.vue

cart.vue
```

this will generate the routes according to nuxt file system routing.

so the user logs in, pings the navigation endpoint and gets something like this:

```
{
    "/einkaufswagen":
        {
            "controller": "cart",
            "title": "Einkaufswagen"
            ....
        },
        ...
}
```

What i would like to do, is to add another route to the router or t modify existing routes to match my needs. This seems to be an edge case that Nuxt didnt consider.

i am now having a very ugly workaround with a "catch all" route and a `<component>` that will then look up a map and return the component... but this also means that the components lose their "page"-properties and also if you are trying to have sub-pages this seems unpossible to tackle.

[I already discussed this here](https://github.com/nuxt/nuxt.js/discussions/7941)

I also found [this smart guy](https://logaretm.com/blog/language-aware-nuxt-routing/) that changes the Vue Router prototypes push function in order to add the locale to each route automatically. So yesterday i tried to karatechop the Router and modify the function that resolves the paths to a certain component, but apparently my brain refuses to help me **ONCE** and i failed miserably 

however, this has become my search for the holy grail

**Comment by MeonisRP** [comment]
I just have the same issue. I made the entire big project on Nuxt and then realised this routing issue. Definitely something should be done here. Not everyone can rebuild their backend to fit Nuxt's router requirements. And it will affect SEO heavily. Especially if a site does already have some url structure to work with

**Comment by saminton** [comment]
I'm having the exact same problem. In our case we're updating an existing site and are reworking the front-end + adding an app so we thought Nuxt would be the perfect solution so as to use the same technology for both. Seeing as the CMS contents already exists and the CMS urls are very flexible we can not rework all the urls to how Nuxt expects them to be.

I've spent the past 2 weeks trying to find a solution to this routing problem and was very surprised to see that Nuxt does not have this functionality built in. It seems rather essentiel to me to be able to control urls from the back-end rather than defining them on the front-end. Has anyone found a solution other than using components for each page ?

Isn't there a problem with the `head` title & meta tags when using components as opposed to pages ?

**Comment by gianpesto** [comment]
> I'm having the exact same problem. In our case we're updating an existing site and are reworking the front-end + adding an app so we thought Nuxt would be the perfect solution so as to use the same technology for both. Seeing as the CMS contents already exists and the CMS urls are very flexible we can not rework all the urls to how Nuxt expects them to be.
> 
> I've spent the past 2 weeks trying to find a solution to this routing problem and was very surprised to see that Nuxt does not have this functionality built in. It seems rather essentiel to me to be able to control urls from the back-end rather than defining them on the front-end. Has anyone found a solution other than using components for each page ?
> 
> Isn't there a problem with the `head` title & meta tags when using components as opposed to pages ?

Eventually i opted for using a catch-all route and dynamic components. yes, you lose all page-abilities on the dynamic components, but you can work around that by using the head and meta properties on the catch-all page.

i have a store that manages all routes that i cannot add to the router, then i search for it on the catchall page and return the right component.

**Comment by saminton** [comment]
Ok thanks, I guess I'll have to make do with what I can. Shame to see that this problem hasn't been addressed 2 years after the original post.

**Comment by gianpesto** [comment]
> Ok thanks, I guess I'll have to make do with what I can. Shame to see that this problem hasn't been addressed 2 years after the original post.

yes, but to their defense, they were working on Nuxt 3, and if i'm not mistaken, this is a feature that comes with the new version. As soon as it runs stable i would consider migrating

**Comment by saminton** [comment]
I'm using Nuxt 3, but haven't found any such functionality. Have you got any links ? Their doc is a bit scarce at the moment.

**Comment by gianpesto** [comment]
> I'm using Nuxt 3, but haven't found any such functionality. Have you got any links ? Their doc is a bit scarce at the moment.

search for _.addRoute_ [on this page](https://v3.nuxtjs.org/api/composables/use-router/)

**Comment by matt-clegg** [comment]
@saminton Did you manage to pass any data into the template component's props with that method?

**Comment by milelazar** [comment]
@saminton addRoute solution works great. Thanks.

Not sure if this is only me, but vue router is loaded before nuxt plugins so on app load you get a router warning "No match found for location with path..."

Route is added and everything works fine when navigating around. Warning is there only on initial load/reload. Any way to get rid of that warning? Or this warning can be ignored

If you hardcode the same route in app/router.options file, warning is gone. But of course this is not a solution.

**Comment by saminton** [comment]
@milelazar I have the same warning since updating to Nuxt 3.0.0. When working with the prerelease versions it wasn't there however. I haven't had the time to look into it and have been ignoring it thus far, but everything seems to work OK for the moment.

**Comment by CernyMatej** [comment]
Is this possible in Nuxt 3 when there's only the `2.x` tag here?

**Comment by manniL** [comment]
The `2.x` label means that it is an issue which was created in the "original Nuxt 2 repo"

**Comment by efimatica** [comment]
> I finally got it working, for anyone interested here's the code that you can use as a plugin. as @gianpesto pointed out this only works with Nuxt 3 as I don't think the `addRoute` function existed before.
> 
> ```js
> export default defineNuxtPlugin(async (nuxtApp) => {
>    
>    const config = useRuntimeConfig();
>    const route = useRoute();
>    const router = useRouter()
>    
>    // Get current page info and template from CMS
> 
>    const { data } = await axios.get(config.public.BASE_API_URL+'page?path='+route.path)
>    
>    // Remove all predefined routes to prevent conflicts
>    
>    router.getRoutes().forEach((route, i)=>{
>       router.removeRoute(route.name)
>    })
>    
>    // Add current page to routes
>    
>    if(data && data.url) {
>       
>       router.addRoute({
>          name: data.id,
>          path: route.path,
>          component: () => import(`../pages/${data.template}.vue`)
>       })
>       
>    }
>    
> })
> ```
> 
> EDIT
> 
> I would not recommend to previous method as the router will not transition to other pages and will instead reload the whole app.
> 
> The solution I ended up using was a mix of built-in dynamic routes from Nuxt for my different post types (as all posts of a certain type will use the same template) and dynamically creating routes for all other pages (where the user can select the associated template in the CMS). The final code looks like this:
> 
> ```js
> export default defineNuxtPlugin(async () => {
>    
>    const router = useRouter()
>    const config = useRuntimeConfig()
>    
>    // Get all page urls + post types with associated templates from Wordpress
>    
>    const { data } = await useApi("vl/routes");
>    
>    // Remove all predefined routes
>    
>    router.getRoutes().forEach((route, i)=>{
>       router.removeRoute(route.name)
>    })
>    
>    // Create routes for each post types
>    
>    data.value.post_types.forEach(item => {
>       router.addRoute({
>          path: `${item.uri}/:slug`,
>          component: () => import(`../pages/${item.template}.vue`)
>       })
>    });
>    
>    // Create routes for pages
>    
>    data.value.pages.forEach(item => {
>       
>       const uri = item.uri ?? '/'
>       const template = item.template ?? '404'
>       
>       router.addRoute({
>          path: uri,
>          component: () => import(`../pages/${template}.vue`)
>       })
>       
>    })
>    
> })
> ```
> 
> With my custom route returning data looking something like this
> 
> ```json
> {
>    "post_types": [
>       {
>          "uri": "/article",
>          "template": "article"
>       },
>       {
>          "uri": "/project",
>          "template": "project"
>       }
>    ],
>    "pages": [
>       {
>          "uri": "/",
>          "template": "home"
>       },
>       {
>          "uri": "/about",
>          "template": "about"
>       }
>    ]
> }
> ```

Hi, ok, runs right, but if in the endpoint has new routes, how it can reload in runtime?
Regards

**Comment by SergkeiM** [comment]
> I finally got it working, for anyone interested here's the code that you can use as a plugin. as @gianpesto pointed out this only works with Nuxt 3 as I don't think the `addRoute` function existed before.
> 
> ```js
> export default defineNuxtPlugin(async (nuxtApp) => {
>    
>    const config = useRuntimeConfig();
>    const route = useRoute();
>    const router = useRouter()
>    
>    // Get current page info and template from CMS
> 
>    const { data } = await axios.get(config.public.BASE_API_URL+'page?path='+route.path)
>    
>    // Remove all predefined routes to prevent conflicts
>    
>    router.getRoutes().forEach((route, i)=>{
>       router.removeRoute(route.name)
>    })
>    
>    // Add current page to routes
>    
>    if(data && data.url) {
>       
>       router.addRoute({
>          name: data.id,
>          path: route.path,
>          component: () => import(`../pages/${data.template}.vue`)
>       })
>       
>    }
>    
> })
> ```
> 
> EDIT
> 
> I would not recommend to previous method as the router will not transition to other pages and will instead reload the whole app.
> 
> The solution I ended up using was a mix of built-in dynamic routes from Nuxt for my different post types (as all posts of a certain type will use the same template) and dynamically creating routes for all other pages (where the user can select the associated template in the CMS). The final code looks like this:
> 
> ```js
> export default defineNuxtPlugin(async () => {
>    
>    const router = useRouter()
>    const config = useRuntimeConfig()
>    
>    // Get all page urls + post types with associated templates from Wordpress
>    
>    const { data } = await useApi("vl/routes");
>    
>    // Remove all predefined routes
>    
>    router.getRoutes().forEach((route, i)=>{
>       router.removeRoute(route.name)
>    })
>    
>    // Create routes for each post types
>    
>    data.value.post_types.forEach(item => {
>       router.addRoute({
>          path: `${item.uri}/:slug`,
>          component: () => import(`../pages/${item.template}.vue`)
>       })
>    });
>    
>    // Create routes for pages
>    
>    data.value.pages.forEach(item => {
>       
>       const uri = item.uri ?? '/'
>       const template = item.template ?? '404'
>       
>       router.addRoute({
>          path: uri,
>          component: () => import(`../pages/${template}.vue`)
>       })
>       
>    })
>    
> })
> ```
> 
> With my custom route returning data looking something like this
> 
> ```json
> {
>    "post_types": [
>       {
>          "uri": "/article",
>          "template": "article"
>       },
>       {
>          "uri": "/project",
>          "template": "project"
>       }
>    ],
>    "pages": [
>       {
>          "uri": "/",
>          "template": "home"
>       },
>       {
>          "uri": "/about",
>          "template": "about"
>       }
>    ]
> }
> ```

This works, and the page loads, but I get a warning in console:

```cli
[Vue Router warn]: No match found for location with path "/"
```
```cli
**ssr:warn** [Vue Router warn]: No match found for location with path "/" 
```

But the pages are rendered and navigation works fine, but not sure why the warning is coming.

**Comment by CernyMatej** [comment]
The warning is present because Vue router doesn't know about the route when the link is rendered on the page (only after navigation).
I think these warnings are rather unfortunate because we cannot turn them off. 😕

What I did to get rid of them is define a catch-all route in the pages directory `[…slug].vue` which just throws a 404 nuxt error.
Seems to get the job done fine, but I'd like a better solution.

**Comment by danielroe** [comment]
We are approaching the Nuxt 2 EOL date (June 30, 2024) - see [this article](https://nuxt.com/blog/nuxt2-eol) for more information. I'm closing this issue as it's marked as a _Nuxt 2 related enhancement_ and it's not a critical issue.

That doesn't mean it might not be relevant for Nuxt 3. If it is, please feel free to open a new issue (or just comment, and I can reopen it). 🙏

Thank you for your understanding and for your contribution to Nuxt! 🎉

**Comment by CernyMatej** [comment]
I think this is absolutely relevant for Nuxt 3. It is actually something I've had to implement myself and I think it would be nice to have support for this in Nuxt itself.
It might be a good idea to create a new issue, though.
I'll go through the topics mentioned here and write something up in the next few days. I'll link it here once the issue is up.

**Comment by danielroe** [comment]
That would be great. I will add that this is certainly possible in Nuxt 3 - in multiple different ways, so this may just be about documenting that for people for whom it's relevant.

**Comment by CernyMatej** [comment]
Linking: https://github.com/nuxt/nuxt/issues/27643
